### PR TITLE
Expose GEOS coverage simplify as a processing algorithm

### PR DESCRIPTION
**Pull request by nyalldawson**
Exposes the new (in geos v12) coverage simplification tool as a processing algorithm.

Open questions:

- should we force validation of the coverage prior to simplification? It's added slowness, but the algorithm does requires valid coverages.
- Should the existing geometry validation check apply to coverages also? Ie we check both the feature geometry AND the coverage geometry?

https://github.com/qgis/QGIS/assets/1829991/fd37d7ea-bcc8-462f-9b4f-cac463e893ad

### CODE REVIEW

**Comment by pathmapper** [comment]
> should we force validation of the coverage prior to simplification? 

+1

> Should the existing geometry validation check apply to coverages also? Ie we check both the feature geometry AND the coverage geometry?

-1

I'm afraid of a failing test because of the coverage for use cases where only a valid geometry is needed...

What about a dedicated coverage validation option for the _Check validity_ processing tool?

File: src/analysis/processing/qgsalgorithmcoveragesimplify.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    addParameter( new QgsProcessingParameterDistance( QStringLiteral( "TOLERANCE" ),
+                                                    QObject::tr( "Tolerance" ), 1.0, QStringLiteral( "INPUT" ), false, 0, 10000000.0) );
+    addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "PRESERVE_BOUNDARY" ), QObject::tr( "Preserve boundary" ), false ) );
+
+    addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Simplified" ) ) );
+}
+
+QString QgsCoverageSimplifyAlgorithm::shortHelpString() const
+{
+    return QObject::tr( "" );
```
Comment (alexbruy) [inline src/analysis/processing/qgsalgorithmcoveragesimplify.cpp:63]: Is this intentional?

File: src/analysis/processing/qgsalgorithmcoveragesimplify.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+    // we have no choice but to build up a list of features in advance
+    QVector< QgsFeature > featuresWithGeom;
+    QVector< QgsFeature > featuresWithoutGeom;
+    QgsGeometryCollection collection;
+
+    const long count = source->featureCount();
+    if ( count >  0 )
+    {
+    featuresWithGeom.reserve( count );
```
Comment (alexbruy) [inline src/analysis/processing/qgsalgorithmcoveragesimplify.cpp:93]: Seems indentation is not correct here.

File: src/analysis/processing/qgsnativealgorithms.cpp [hunk 3]
```diff
@@ -47,6 +47,7 @@
 #include "qgsalgorithmconstantraster.h"
 #include "qgsalgorithmconverttocurves.h"
 #include "qgsalgorithmconvexhull.h"
+#include "qgsalgorithmcoveragesimplify.h"
```
Comment (alexbruy) [inline src/analysis/processing/qgsnativealgorithms.cpp:50]: Probably this should be put inside `ifdef` as algorithm requires GEOS >=3.12

File: src/analysis/processing/qgsnativealgorithms.cpp [hunk 4]
```diff
@@ -313,6 +314,7 @@ void QgsNativeAlgorithms::loadAlgorithms()
   addAlgorithm( new QgsConstantRasterAlgorithm() );
   addAlgorithm( new QgsConvertToCurvesAlgorithm() );
   addAlgorithm( new QgsConvexHullAlgorithm() );
+  addAlgorithm( new QgsCoverageSimplifyAlgorithm() );
```
Comment (alexbruy) [inline src/analysis/processing/qgsnativealgorithms.cpp:317]: Same as above, maybe `ifdef` this?

**Comment by nyalldawson** [comment]
@pathmapper 

> I'm afraid of a failing check because of the coverage for use cases where only a valid geometry is needed...

Ah, I think you misunderstood. I was referring specifically to this algorithm only, making it do a forced coverage validity check if the geometry check setting is anything other then "ignore". Other non-coverage related algorithms wouldn't be changed.

> What about a dedicated coverage validation option for the Check validity processing tool?

It's on its way! (I just cant get the geos coverage check method to work correctly right now 😳)

File: src/analysis/processing/qgsnativealgorithms.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/analysis/processing/qgsnativealgorithms.cpp:50]: I think the "right" approach is that we still expose the algorithm and only catch the error when it's actually run. This means models shared with users with older geos won't be completely broken, but rather they'll get an explanatory error when they try to run them only.

But happy for second opinions!

**Comment by pathmapper** [comment]
Thanks, got it :-) Then I'm +1 to check both the feature geometry AND the coverage geometry if the geometry check setting is anything other then "ignore".

> It's on its way! 

:+1:

File: src/analysis/processing/qgsalgorithmcoveragevalidate.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  try
+  {
+    result = geos.validateCoverage( gapWidth, &invalidEdges, &error ) ;
+  }
+  catch ( QgsNotSupportedException &e )
+  {
+    throw QgsProcessingException( e.what() );
+  }
+
+  if ( result == Qgis::CoverageValidityResult::Error )
```
Comment (alexbruy) [inline src/analysis/processing/qgsalgorithmcoveragevalidate.cpp:131]: Looks like leftover and can be removed

File: src/analysis/processing/qgsnativealgorithms.cpp (hunk 3 above)
Comment (alexbruy) [reply src/analysis/processing/qgsnativealgorithms.cpp:50]: No, it is totally fine from this point of view.

File: src/analysis/processing/qgsalgorithmcoveragesimplify.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmcoveragesimplify.cpp]: It was a "todo" :laughing:

**Pull request merged**